### ISSUE TITLE
Plug-in looses accessory state on reload

### ISSUE THREAD

**Issue by azinman** [kuestess/homebridge-platform-insteonlocal#160]
Every time I restart Homebridge, all my Insteon devices forget their room assignments made in the Home app and all go to the default room. Similarly, I believe Siri Shortcuts also fail to work without modifications. I’m guessing it’s ignoring the accessory cache and rebuilding state from scratch? What’s confusing to me about that is my config.json outlined all the devices, so it’s not like they need to be discovered. On boot it says it finds them in the config and then proceeds to initialize each one as a platform accessory.

**Comment by kuestess** [comment]
@azinman Do you by chance have an older iPhone/iPad connected to your home data?

**Comment by azinman** [comment]
Can you qualify older? I have many devices.  Looking at the code, I don’t see anything that would reload the mapping on startup. It seems to rediscover the devices and then put them into homebridge, unless I read that wrong. Homebridge has a cache of discovered devices with a callback at boot which I believe is being ignored.

**Comment by kuestess** [comment]
@azinman You're correct that I don't do anything with the cache, although it hasn't caused any problems like you're describing.  A quick search showed a couple threads where older (iOS 10 vintage, or around there) caused more contemporary devices to lose the room assignments.

**Comment by azinman** [comment]
I don’t believe I have any such devices in my house. If you don’t use the cache, then what keeps the room assignments? I don’t see it in the Insteon.json either

**Comment by kuestess** [comment]
@azinman I believe all of the room assignments are handled by Homekit rather than home bridge.  Accessories are designated by uuid, which the plugin generates baed upon the device name.  Let me look at the cache code from the plugin template to see if there's anything there that might give a hint.

**Comment by azinman** [comment]
From going over the Homebridge docs, it seems their cache retains all of this information. If you don't use their cache, then I believe such information will be lost when Homebridge starts over again and you dynamically pull in all of the accessories from Insteon, even if their uuid remains the same.

From https://github.com/homebridge/homebridge/issues/1614:
"On startup, homebridge loads these, before starting its web server. If you delete this file (after shutting down homebridge), homebridge will start without exposing these accessories, causing HomeKit to delete them (incl. any associations to rooms, groups, scenes, and automations). If the plugin re-creates the accessories, HomeKit will see them as new accessories, in the default room (or rather; in the room with the homebridge bridge accessory)."

**Comment by azinman** [comment]
This also explains how restarting homebridge causes my scenes with Insteon devices to no longer work.

**Comment by kuestess** [comment]
@azinman I've only seen that behavior when I completely change Hub/PLM (many times for testing) - HomeKit forgets the devices and my HomeKit scenes no longer work.  I don't see that when I restart Homebridge -  devices are still in their respective rooms and scenes work as expected.  Have you looked at permissions on your `.homebridge` folder?

**Comment by azinman** [comment]
Everything in /var/lib/homebridge is homebridge:homebridge. I have other plugins like nest and ring that do not lose their room associations during homebridge restart. Only the Insteon plugin in my setup has this flaw, and I strongly believe that's because of the lack of using the cached accessory API. I could try to fix this on my own branch by using the cache but unfortunately my employer won't let me work on open source code, so I wouldn't be able to publish my patch.

It's unclear to me how it could work for you -- maybe the details are in the setup. I have a lot of Insteon devices. I do have an Insteon hub, but I use the plugin directly with the Insteon PLM.

**Comment by kuestess** [comment]
@azinman Did some looking into the accessory cache and it seems that this is only for newer, dynamic platforms where insteonlocal is an (older) static platform.  The issue appears to occur when you have a mix of static/dynamic platforms - room assignments for static platforms get nuked.  Refactoring the plugin will take some time, more than the quick add I expected :-).  Just want to be transparent about timing!

**Comment by azinman** [comment]
Ahh. I do have both (about 7 plugins), of which at least my own is for sure a dynamic. Thank you for doing this. Much much appreciated! Happy to alpha test

**Comment by azinman** [comment]
@kuestess Any luck with progress?

**Comment by kuestess** [comment]
@azinman Got it working with just lightbulb/dimmers, and then decided to add other devices.  Unfortunately, that created a snowball of method dependencies which I'm working through now.  Should have something soon.  Stay tuned,

**Comment by kuestess** [comment]
@azinman Just uploaded very alpha code to the `dev` branch.  Limited testing, but lights/scenes seem to work.  Still pretty messy, but like to see what you find.

**Comment by azinman** [comment]
@kuestess This is working for me! I just have fans & switches. I haven't tested scenes just yet. Thank you so much!

**Comment by kuestess** [comment]
@azinman Glad its working for you.  I hadn't done much in Typescript before so it was interesting....made me love and hate Typescript.

**Comment by azinman** [comment]
Ship it! :)

What is the rest of the bridge written in then?

**Comment by kuestess** [comment]
@azinman Its Javascript.  Just not used to the strict typing on Typescript :-)

**Comment by azinman** [comment]
Hopefully this doesn't get marked as stale and go away. @kuestess any plans to merge this in? It seems to disable the custom express-server-UI, which I'd love to get back to manually link a couple new devices.

**Comment by kuestess** [comment]
@azinman Yes, this is probably next on my list - would love to get migrated over and get everything fully functional, including the ui.  Just need to find the time to dig back in.  Still seem to work for you?  I had some devices 'not responding' that popped up, but not sure yet what the cause was.

**Comment by azinman** [comment]
I have devices “not responding” after updating them in the Home app. I wasn’t sure that it was related to this change or something else. I feel like that happened after I updated Homebridge, but maybe not?

**Comment by azinman** [comment]
Any updates @kuestess ? I’d hate to see this closed. I wish I could help with code but my employer won’t let me :(

**Comment by kuestess** [comment]
@azinman No plans to close.  I've been working on the `dev` branch and plan to merge once I get time to polish it up.  Need to get a few issues on the current branch resolved as well, but the plan is to make this a dynamic platform.

**Comment by azinman** [comment]
@kuestess Good luck on the refactor. Not sure if you’ve seen but Insteon’s servers are now down, and it seems the company is dead. This plug-in is what’s keeping my existing setup now alive! It has some issues with the previous dev branch I tried (namely devices “aren’t responding”), and is missing the linking web UI that I’ll now have to rely on. So let me be a little cheerleader encouraging you!

**Comment by kuestess** [comment]
@azinman Yep, the repo and reddit have been blowing up - glad its keeping you afloat.  Merging the dev branch into master is my next stop, assuming no deal-breaker bugs.  Thank you for the encouragement ;-).

**Comment by azinman** [comment]
I assume you have local changes you haven’t pushed to dev? Because the dev branch right now hasn’t been updated in a long time and has some issues with non responsive devices. You also mentioned you were working on it. I’m happy to give it a spin and let you know — I have dozens of devices to try it with.

**Comment by coreTech51** [comment]
Hi,

With Insteon servers down, will the Hub still function locally? I can’t get nothing to work. I’m getting this in my logs.
Thanks.

4/17/2022, 10:35:46 AMPlatform Insteonlocal BridgeLoaded plugin 'homebridge-platform-insteonlocal'
4/17/2022, 10:35:46 AMPlatform Insteonlocal BridgeLoading 1 platforms...
4/17/2022, 10:35:46 AMPlatform Insteonlocal BridgeInsteon Local PlatormStarted connection watcher...
4/17/2022, 10:35:46 AMPlatform Insteonlocal BridgeInsteon Local PlatormFound 92 devices in config
4/17/2022, 10:35:47 AMPlatform Insteonlocal BridgeERROR(node:28247) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 connect listeners added to [Insteon]. Use emitter.setMaxListeners() to increase limit
4/17/2022, 10:35:47 AMPlatform Insteonlocal BridgeERROR(Use `node --trace-warnings ...` to show where the warning was created)
4/17/2022, 10:35:47 AMPlatform Insteonlocal BridgeBridge is running on port 51886.
4/17/2022, 10:35:47 AMPlatform Insteonlocal BridgeWARNINGunhandled rejection: Error: Status code expected 200, got 401

Sent from my iPad
Aaron Harvey
***@***.***

---------------------------------------------
Confidential Notice: This communication, including attachments, may contain information that is confidential and protected by the attorney/client or other privileges. It constitutes non-public information intended to be conveyed only to the designated recipient(s). If reader or recipient of this communication is not the intended recipient, an employee or agent of the intended recipient who is responsible for delivering it to the intended recipient, or you believe that you have received this communication in error, please notify the sender immediately by return email and promptly delete this email, including attachments without reading or saving them in any manner. The unauthorized use, dissemination, distribution, or reproduction of this email, including attachments, is prohibited and may be unlawful. Receipt by anyone other than the intended recipient(s) is not a waiver of any attorney/client or other privilege.
---------------------------------------------

**Comment by kuestess** [comment]
@azinman Yes, have local changes that I haven't pushed yet.  Working on getting the UI part ported to typescript.

**Comment by kuestess** [comment]
@azinman Sorry for the delay in any update - things are finally (sort of) starting to calm down.  Just pushed a new version to the dev branch that adds the eventListener and should fix the 'no response' issues.  Should also be at parity with fixes implemented in the master branch.  Next stop is getting the ui ported over to typescript.
